Add 'Install the browser extension' alert + animation

tjkandala opened the pull request:
Implements #14196.

TODO:
- [x] - Add alert
- [x] - CSS for animations
- [x] - Display different alert depending on browser
- [x] - Integration tests
- [x] - Snapshots for new components

tjkandala commented on web/src/Layout.tsx:
```diff
@@ -160,6 +160,12 @@ export const Layout: React.FunctionComponent<LayoutProps> = props => {
     const breadcrumbProps = useBreadcrumbs()
 
+    // TODO description
+    const startUserNavExtensionAnimationReference = useRef<null | (() => void)>(null)
```
I need to find a much shorter name for this

tjkandala replied:
I'm also not sure if a ref is the best solution to passing this function up to `Layout` so that it can be called from `RepoContainer`. However, I strongly prefer this to handling the animation state in `Layout` and drilling multiple props down to `UserNavItem`. Could we try dispatching a custom event in `RepoContainer` + listening for it in `UserNavItem`?

felixfbecker replied:
A ref does seem very hacky. Events in React are just callbacks passed through props. If there is a lot of logic that you don't want to move to `Layout`, I'd put that in a hook that you export from the user nav menu folder, so the Layout component just has to use that hook and the logic is colocated with where it's used. But also see my comment below on potentially reducing the logic to a minimum through CSS animations.

tjkandala commented on web/src/nav/UserNavItem.tsx:
```diff
@@ -1,166 +1,250 @@
+    const startUserNavExtensionAnimation = useCallback(async () => {
+        if (!isAnimating) {
+            setIsAnimating(true)
+            // 100ms delay
+            await new Promise(resolve =>
+                delayManager.setTimeout(() => {
+                    // tooltip eases out (400ms [in CSS])
+                    setShowTooltip(true)
+                    console.log('show tooltip')
+
+                    resolve()
+                }, 100)
+            )
+            // 50ms delay
+            await new Promise(resolve =>
+                delayManager.setTimeout(() => {
```
Is this our preferred approach for serial timeouts? Since there are only three timeouts, we could nest them as well, but that might be just as ugly.

felixfbecker replied:
I think this approach is fine, but it seems like a lot of JS for a simple animation sequence. Wouldn't this be a lot cleaner with CSS animations? To start it, you'd just assign a CSS class, and the rest (revealing and hiding elements) follows in CSS using keyframes.

tjkandala commented:
![extanim](https://user-images.githubusercontent.com/37420160/95157055-0234b780-0766-11eb-93f0-d20444b17b68.gif)

I added a small timeout after the 5th hover to prevent layout shift before the user has a chance to read hover content. 

@AlicjaSuska I temporarily implemented only the timing portion of the background animation for user avatars. How should the animation look when the avatar isn't set?

I noticed that it took a lot of steps to enable the popover after it's dismissed, so I added some functions to the `Window` object to make this PR easier to play with:
- `onExtensionAlertDismissed`: Triggers the tooltip + avatar animation
- `showExtensionAlert`: Displays alert

felixfbecker commented:
That layout shift seems bad, I'm worried that it would cause the alert to be perceived as annoying and be clicked away. This of course was only caused by the change to count the hovers instead of just counting days, where it would only appear on page load. I think we can have the best of both worlds by not showing the alert immediately, but only storing the hover count and only showing the alert on a page load/initial render.

tjkandala commented:
<details open>
<summary>Animation w/ default icon (no profile picture set)</summary>
<img src="https://user-images.githubusercontent.com/37420160/95266194-b6831c00-0800-11eb-84d8-eb176cfe3b10.gif" />
</details>

<details>
<summary>Animation w/ profile picture</summary>
<img src="https://user-images.githubusercontent.com/37420160/95266190-b4b95880-0800-11eb-9ee7-1e29ae46196a.gif" />
</details>

tjkandala commented on web/src/repo/RepoContainer.tsx:
```diff
@@ -335,15 +355,28 @@ export const RepoContainer: React.FunctionComponent<RepoContainerProps> = props
         localStorage.setItem(HOVER_COUNT_KEY, count.toString(10))
     }, [])
 
+    // DEBUG
```
Will remove after design review

tjkandala commented on web/src/Layout.tsx:
```diff
@@ -160,6 +161,19 @@ export const Layout: React.FunctionComponent<LayoutProps> = props => {
+        startExtensionAlertAnimation()
+    }, [startExtensionAlertAnimation])
+
+    // DEBUG
+    useEffect(() => {
+        // eslint-disable-next-line
+        ;(window as any).onExtensionAlertDismissed = onExtensionAlertDismissed
+    }, [onExtensionAlertDismissed])
+
```
Will remove after design review

tjkandala commented on web/src/integration/blob-viewer.test.ts:
```diff
@@ -259,9 +262,42 @@ describe('Blob viewer', () => {
-            it.skip(`shows an alert about the browser extension when the user has seen ${HOVER_THRESHOLD} hovers`, async () => {
+            it(`shows an alert about the browser extension when the user has seen ${HOVER_THRESHOLD} hovers`, async () => {
                 await driver.page.goto(`${driver.sourcegraphBaseUrl}/github.com/sourcegraph/test/-/blob/test.ts`)
-                // TODO
+                await driver.page.evaluate(HOVER_COUNT_KEY => localStorage.removeItem(HOVER_COUNT_KEY), HOVER_COUNT_KEY)
+                await driver.page.reload()
```
Is there a better way to clear `localStorage` before each test? Clearing it in `page.evaluateOnNewDocument` would break this test, since that would run on reload. There may be a solution with request interception. 

Fwiw, the test still passed every time I ran it without reloading after clearing `localStorage`. I just want to eliminate any possibility of flakiness.

felixfbecker replied:
This is all I can think of - would ensure 100% reproducability:

```ts
beforeEach(async () => {
  await driver.page.evaluate(() => localStorage.clear())
})
```

AlicjaSuska commented:
@tjkandala
- I've added a dark alert variant for the [animation background](https://www.figma.com/file/V5t7KOzC6PIX75L4dz8IFc/RFC-221-Browser-extension-improve-discoverability-and-awareness-Approved?node-id=869%3A0) and for the [avatar background](https://www.figma.com/file/V5t7KOzC6PIX75L4dz8IFc/RFC-221-Browser-extension-improve-discoverability-and-awareness-Approved?node-id=869%3A47)
- when it comes to the user name animation. We can flash the [background behind the text](https://www.figma.com/file/V5t7KOzC6PIX75L4dz8IFc/RFC-221-Browser-extension-improve-discoverability-and-awareness-Approved?node-id=869%3A67). Please, do no flash the text itself. Also, if you're running out of time, it is something we can cut out of the scope :)

felixfbecker commented:
@tjkandala the build is broken on this, could you fix?
